go/types/type_cache.go

```diff
@@ -539,11 +539,12 @@ func MakeStructType(name string, fieldNames []string, fieldTypes []*Type) *Type
 }
 
 func MakeUnionType(elemTypes ...*Type) *Type {
-	staticTypeCache.Lock()
```

Comment:
Putting the lock here was too coarse because sometimes makeSimplifiedUnion tries to acquire it again re-entrantly.